### PR TITLE
refactor(connection): refactor mqtt/edgex connection share

### PR DESCRIPTION
**Pull request by SuperRxan**
Signed-off-by: Jianxiang Ran <rxan_embedded@163.com>

### CODE REVIEW

File: internal/topo/connection/clients/mqtt/mqtt_wrapper.go [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+				consumer.ConsumerChan <- types.MessageEnvelope{
+					MqttMsg: message,
+				}
+			}
+		}
+	}
+}
+
+func (mc *mqttClientWrapper) Publish(c api.StreamContext, topic string, message []byte) error {
+	subId := fmt.Sprintf("%s_%s_%d", c.GetRuleId(), c.GetOpId(), c.GetInstanceId())
```
Comment (ngjaying) [inline internal/topo/connection/clients/mqtt/mqtt_wrapper.go:131]: Is this needed? Do not do such calculation in a very high frequency op

File: internal/topo/connection/clients/mqtt/mqtt_wrapper.go [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+	return nil
+}
+
+func (mc *mqttClientWrapper) Subscribe(c api.StreamContext, subChan []types.TopicChannel, messageErrors chan error) error {
+	log := c.GetLogger()
+
+	mc.subLock.Lock()
+	defer mc.subLock.Unlock()
+
+	subId := fmt.Sprintf("%s_%s_%d", c.GetRuleId(), c.GetOpId(), c.GetInstanceId())
```
Comment (ngjaying) [inline internal/topo/connection/clients/mqtt/mqtt_wrapper.go:150]: Is it possible to calculated beforehand and save?

Comment (ngjaying) [reply internal/topo/connection/clients/mqtt/mqtt_wrapper.go:150]: If this is only run once, then it is ok

File: internal/topo/connection/clients/mqtt/mqtt_wrapper.go [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+
+	subTopics := clients.SubscribedTopics{
+		Topics: make([]string, 0),
+	}
+
+	for _, tpChan := range subChan {
+		tpc := tpChan.Topic
+		subTopics.Topics = append(subTopics.Topics, tpc)
+		sub, found := mc.topicSubscriptions[tpc]
+		if found {
```
Comment (ngjaying) [inline internal/topo/connection/clients/mqtt/mqtt_wrapper.go:167]: what about qos different

File: internal/topo/connection/clients/edgex/edgex_wrapper.go [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+
+	err = client.Connect()
+	if err != nil {
+		return nil, err
+	}
+	return cliWpr, nil
+}
+
+func (mc *edgexClientWrapper) Publish(c api.StreamContext, topic string, message []byte) error {
+	subId := fmt.Sprintf("%s_%s_%d", c.GetRuleId(), c.GetOpId(), c.GetInstanceId())
```
Comment (ngjaying) [inline internal/topo/connection/clients/edgex/edgex_wrapper.go:104]: do not calculate this everytime

**Review by ngjaying — CHANGES_REQUESTED** [review]
Shall we name these resource?

File: internal/topo/connection/clients/mqtt/mqtt_wrapper.go [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+				con.SubErrors <- token.Error()
+			}
+		}
+	}
+}
+
+func (mc *mqttClientWrapper) newMessageHandler(sub *mqttSubscriptionInfo) pahoMqtt.MessageHandler {
+	return func(client pahoMqtt.Client, message pahoMqtt.Message) {
+		if sub != nil {
+			for _, consumer := range sub.topicConsumers {
```
Comment (ngjaying) [inline internal/topo/connection/clients/mqtt/mqtt_wrapper.go:121]: The broadcast here may be blocked forever?

File: internal/topo/source/mqtt_source.go [hunk 6]
```diff
@@ -111,131 +73,80 @@ func (ms *MQTTSource) Configure(topic string, props map[string]interface{}) erro
...
-			}
-		}
-
-		select {
-		case consumer <- api.NewDefaultSourceTuple(result, meta):
-			log.Debugf("send data to source node")
-		case <-ctx.Done():
-			return
-		}
+	messages := make(chan types.MessageEnvelope)
```
Comment (ngjaying) [inline internal/topo/source/mqtt_source.go:95]: won't be closed

File: internal/topo/connection/clients/mqtt/mqtt_wrapper.go (hunk 3 above)
Comment (SuperRxan) [reply internal/topo/connection/clients/mqtt/mqtt_wrapper.go:170]: will document this

File: pkg/api/stream.go [hunk 7]
```diff
@@ -16,6 +16,8 @@ package api
 
 import (
 	"context"
+	MQTT "github.com/eclipse/paho.mqtt.golang"
+	"github.com/edgexfoundry/go-mod-messaging/v2/pkg/types"
```
Comment (ngjaying) [inline pkg/api/stream.go:20]: Don't import these implementations in the api package

File: pkg/api/stream.go [hunk 8]
```diff
@@ -188,3 +189,26 @@ const (
...
 
 type Qos int
+
+type MessageClient interface {
+	Subscribe(c StreamContext, subChan []TopicChannel, messageErrors chan error) error
+	Release(c StreamContext)
+	Publish(c StreamContext, topic string, message []byte) error
+}
+
+type MessageEnvelope struct {
```
Comment (ngjaying) [inline pkg/api/stream.go:199]: Don't put these structs?

File: internal/topo/connection/factory/buildin.go [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package factory
+
+import (
+	"github.com/lf-edge/ekuiper/internal/topo/connection/clients"
+	"github.com/lf-edge/ekuiper/internal/topo/connection/clients/mqtt"
+)
+
```
Comment (ngjaying) [inline internal/topo/connection/factory/buildin.go:21]: filename should be builtin

File: internal/topo/connection/clients/mqtt/mqtt_wrapper.go [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+				select {
+				case consumer.ConsumerChan <- &api.MessageEnvelope{MqttMsg: message}:
+					break
+				default:
```
Comment (ngjaying) [inline internal/topo/connection/clients/mqtt/mqtt_wrapper.go:127]: Print a log here

File: internal/topo/connection/clients/edgex/edgex_wrapper.go [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+					}
+					return
+				}
+				//broadcast to all topic subscribers
+				if sub != nil {
+					for _, consumer := range sub.topicConsumers {
+						select {
+						case consumer.ConsumerChan <- &api.MessageEnvelope{EdgexMsg: msg}:
+							break
+						default:
```
Comment (ngjaying) [inline internal/topo/connection/clients/edgex/edgex_wrapper.go:143]: print some log here

**Review by ngjaying — CHANGES_REQUESTED** [review]
Some small problems

**Review by ngjaying — APPROVED** [review]

**Pull request merged**